transaction.js

```diff
@@ -0,0 +1,146 @@
+ * @param {MongoClient} client A MongoClient that is connected to a cluster with the sample_airbnb database
+ * @param {String} userEmail The email address of the user who is creating the reservation
+ * @param {String} nameOfListing The name of the Airbnb listing to be reserved
+ * @param {Array.Date} reservationDates An array of the date(s) for the reservation
+ * @param {Object} reservationDetails An object containing additional reservation details that need to be stored with the reservation
+ */
+async function createReservation(client, userEmail, nameOfListing, reservationDates, reservationDetails) {
+
+    /**
+     * The users collection in the sample_airbnb database
+     */
+    const usersCollection = client.db("sample_airbnb").collection("users");
+
+    /**
+     * The listingsAndReviews collection in the sample_airbnb database
+     */
+    const listingsAndReviewsCollection = client.db("sample_airbnb").collection("listingsAndReviews");
+
+    /**
+     * The reservation document that will be added to the users collection document for this user
+     */
+    const reservation = createReservationDocument(nameOfListing, reservationDates, reservationDetails);
+
+    // Step 1: Start a Client Session
+    // See http://bit.ly/Node_startSession for the startSession() docs
+    const session = client.startSession();
+
+    // Step 2: Optional. Define options to use for the transaction
+    const transactionOptions = {
+        readPreference: 'primary',
+        readConcern: { level: 'local' },
+        writeConcern: { w: 'majority' }
+    };
+
+    try {
+        // Step 3: Use withTransaction to start a transaction, execute the callback, and commit (or abort on error)
+        // Note: The callback for withTransaction MUST be async and/or return a Promise.
+        // See http://bit.ly/Node_withTransaction for the withTransaction() docs
+        const transactionResults = await session.withTransaction(async () => {
+
+            // Important:: You must pass the session to each of the operations   
+
+            // Add a reservation to the reservations array for the appropriate document in the users collection
+            const usersUpdateResults = await usersCollection.updateOne(
+                { email: userEmail },
+                { $addToSet: { reservations: reservation } },
+                { session });
+            console.log(`${usersUpdateResults.matchedCount} document(s) found in the users collection with the email address ${userEmail}.`)
+            console.log(`${usersUpdateResults.modifiedCount} document(s) was/were updated to include the reservation.`);
+
+            // Check if the Airbnb listing is already reserved for those dates. If so, abort the transaction.
+            const isListingReservedResults = await listingsAndReviewsCollection.findOne({ name: nameOfListing, datesReserved: { $in: reservationDates } });
```

Comment:
you need to remember to pass `session` into _every_ method that interacts with the server, this looks like its missing it, unless that is intentional.

Reply:
Good catch! Thank you!  I've updated the code.  

I want to make sure I have a full understanding.  My initial inclination was that, since this is a "read" operation and not impacting the data itself, it didn't need the session (there was no concern about "rolling back" the operation if it failed).  However, we want to ensure that the "read" occurs as part of this transaction--if an error occurs while checking the dates, we want to abort the transaction.  Also, we want to ensure that the date checking is part of the transaction to ensure we have the correct version of the data at the point in time that the transaction occurs (for purposes of isolation and concurrency control).  Is that right?

Reply:
Exactly, you want to have "causal consistency" here (read: Read your Writes). So when you make this read, you want to ensure its done the context of the transaction so you can ensure you get the expected result.

Reply:
Got it. Thank you!